File: mapeable/Mapeable/Comparaciones/ComparadorDeUnaPropiedad.vb
Imports System.Reflection

Namespace Comparaciones
    Public Class ComparadorDeUnaPropiedad
        Private laPropiedadOrigen As PropertyInfo
        Private elValor As Object
        Private elOtroValor As Object
        Private esIgual As Boolean

        Sub New(laPropiedad As String, elTipoDeLaClase As Type)
            Dim losFiltros As BindingFlags = BindingFlags.Instance Or BindingFlags.Public
            laPropiedadOrigen = elTipoDeLaClase.GetProperty(laPropiedad, losFiltros)
        End Sub

        Function LaPropiedadEsIgual(unObjeto As Object, otroObjeto As Object) As Boolean
            ObtengaElValor(unObjeto)
            ObtengaElOtroValor(otroObjeto)
            CompareLosValores()

            Return esIgual
        End Function

        Private Sub ObtengaElOtroValor(unObjeto As Object)
            elValor = laPropiedadOrigen.GetValue(unObjeto, Nothing)
        End Sub

        Private Sub ObtengaElValor(otroObjeto As Object)
            elOtroValor = laPropiedadOrigen.GetValue(otroObjeto, Nothing)
        End Sub

        Private Sub CompareLosValores()
            Dim elTipo As Type
            elTipo = laPropiedadOrigen.PropertyType

            If elTipo.Equals(GetType(String)) Then
                esIgual = String.Equals(elValor, elOtroValor)
            ElseIf elTipo.IsClass Then
                Dim elComparador As New ComparadorBase()
                esIgual = elComparador.EsIgualQue(elValor, elOtroValor)
            Else
                esIgual = Object.Equals(elValor, elOtroValor)
            End If
        End Sub
    End Class
End Namespace

File: mapeable/Mapeable.ComparacionesParaPruebasUnitarias/Comparaciones/ComparadorDeUnaPropiedad.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Reflection;

namespace Comparaciones
{
    public class ComparadorDeUnaPropiedad
    {
        private PropertyInfo laPropiedadOrigen;
        private object elValor;
        private object elOtroValor;

        private bool esIgual;
        public ComparadorDeUnaPropiedad(string laPropiedad, Type elTipoDeLaClase)
        {
            BindingFlags losFiltros = BindingFlags.Instance | BindingFlags.Public;
            laPropiedadOrigen = elTipoDeLaClase.GetProperty(laPropiedad, losFiltros);
        }

        public bool LaPropiedadEsIgual(object unObjeto, object otroObjeto)
        {
            Type elTipo = unObjeto.GetType();

            if (EsUnEnumerableGenerico(elTipo))
            {
                ComparadorBaseDeColecciones elComparador = new ComparadorBaseDeColecciones();

                IEnumerable<object> estaColeccion = (IEnumerable<object>)unObjeto;
                IEnumerable<object> laOtraColeccion = (IEnumerable<object>)otroObjeto;

                esIgual = elComparador.EsIgualQueLaColeccion(estaColeccion, laOtraColeccion);
            }
            else
            {
                ObtengaElValor(unObjeto);
                ObtengaElOtroValor(otroObjeto);
                CompareLosValores();
            }

            return esIgual;
        }

        private void ObtengaElOtroValor(object unObjeto)
        {
            elValor = laPropiedadOrigen.GetValue(unObjeto, null);
        }

        private void ObtengaElValor(object otroObjeto)
        {
            elOtroValor = laPropiedadOrigen.GetValue(otroObjeto, null);
        }

        private void CompareLosValores()
        {
            Type elTipo = default(Type);
            elTipo = laPropiedadOrigen.PropertyType;

            if (elTipo.Equals(typeof(string)))
            {
                esIgual = string.Equals(elValor, elOtroValor);
            }
            else if (elTipo.IsClass)
            {
                ComparadorBase elComparador = new ComparadorBase();
                esIgual = elComparador.EsIgualQue(elValor, elOtroValor);
            }
            else {
                esIgual = object.Equals(elValor, elOtroValor);
            }
        }

        private bool EsUnEnumerableGenerico(Type elTipo)
        {
            if (typeof(IEnumerable<object>).IsAssignableFrom(elTipo))
                return true;
            else
                return false;
        }
    }
}